mantidproject/mantid
# Project Recovery - Background Saving

**Description of work.**
This PR implements the saving functionality for background saving. The following changes are included:

- Added a function to the Python API to write workspace histories
- Updated `ProjectSerialiser` to skip attempting file loads in recovery mode
- Refactored `ProjectSeraliser` to remove any workspaces not in the project file, in recovery mode
- Added `ProjectRecoveryThread` which contains a background thread that saves out all workspace histories and GUI windows periodically
- These are all saved to `%appdata%/recovery/*timestamp*` a maximum of 5 are kept in rotation
- Saves every minute (configurable)
- `ProjectRecoveryThread` will stop if anything fails to work for any reason.
- Added new configuration keys for all of these options
- Information logging indicates when saving happens

*To Do in a future PR*
- Recovery steps (loading)
- Folder clean-up on clean Mantid exit
- Resolve some window listeners not being correctly linked, so clear all memory leaves GUI windows behind
- Handle `secondsBetween <= 0`

**To test:**
- *Optional* set `projectRecovery.secondsBetween = 60` in the properties file to a lower value (bottom of file)
- Open `%appdata%` folder on your machine
- Open Mantid Plot
- Set debugging to ~notice~ debug
- Check that saving triggers correctly
- Ensure after 5 saves the folders in `recovery` are being rotated
- Create a sample workspace and open a GUI
- Check the saved files in the timestamped folder (project file and .py) contain the new GUI / WS

-----------------------------

- Add a std::exception anywhere in the main thread body
- Check it is caught correctly and the thread stops saving
- Add a std::exception anywhere in the ProjectSerialiser save body
- *same again*
- Add an error to the Python simple API changes (write_all_workspaces_histories)
- *same again*
- Check the configuration key stops project saving when `projectRecovery.enabled` is set to false

-----------------------------

- Open a large number of GUI windows
- Open a large number of workspaces (GenerateMess can be used)
- Check that the save triggering does not cause any freezes or hangs

Does this update require release notes?

Release notes will be updated as part of the loading mechanism in the future PR

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
